src/orange/detail/BIHMaker.hh

```diff
@@ -0,0 +1,116 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHMaker.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <vector>
+
+#include "orange/BoundingBox.hh"
+#include "orange/OrangeData.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Create a bounding interval hierarchy from supplied bounding boxes.
+ *
+ * This implementation matches the structure proposed in the original
+ * paper [1]. Partitioning is done on the basis of bounding box centers using
+ * the "longest dimension" heuristic. All leaf nodes contain either a single
+ * volume id, or multiple volume ids if the volumes have bounding boxes that
+ * share the same center.
+ *
+ * [1] C. Wachter, Carsten and A. Keller, "Instant Ray Tracing: The Bounding
+ * Interval Hierarchy" Eurographics Symposium on Rendering, 2006,
+ * doi:10.2312/EGWR/EGSR06/139-149}
+ */
+class BIHMaker
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecBBox = std::vector<BoundingBox>;
+    using Storage = Collection<LocalVolumeId,
+                               Ownership::value,
+                               MemSpace::host,
+                               OpaqueId<LocalVolumeId>>;
+    using VecNodes = std::vector<BIHNode>;
+    //!@}
+
+  public:
+    // Construct from vector of bounding boxes and storage for LocalVolumeIds
+    explicit CELER_FUNCTION BIHMaker(VecBBox bboxes, Storage* storage);
+
+    // Create BIH Nodes
+    CELER_FUNCTION VecNodes operator()();
+
+  private:
+    /// TYPES ///
+
+    using VecReal3 = std::vector<Real3>;
+    using VecIndices = std::vector<LocalVolumeId>;
+    using PairVecIndices = std::pair<VecIndices, VecIndices>;
+    using VecAxes = std::vector<Axis>;
+    using AxesCenters = std::vector<std::vector<double>>;
+
+    struct Partition
+    {
+        Axis axis = Axis::size_;
+        double location;
+        explicit CELER_FUNCTION operator bool() const
+        {
+            return axis == Axis::x || axis == Axis::y || axis == Axis::y;
+        }
+    };
+
+    //// DATA ////
+
+    VecBBox bboxes_;
+    Storage* storage_;
+
+    //// HELPER FUNCTIONS ////
+
+    // Recursively construct BIH nodes for a vector of bbox indices
+    void construct_tree(VecIndices const& indices, VecNodes& nodes);
+
+    // Find a suitable partition for the given bounding boxes
+    Partition
+    find_partition(VecIndices const& indicies, VecReal3 const& centers);
+
+    // Divide bboxes into left and right branches based on a partition
+    PairVecIndices partition_bboxes(VecIndices const& indices,
+                                    VecReal3 const& centers,
+                                    Partition const& p);
```

Comment:
It would be very useful for testing to have the partitioning scheme be a separate detail class.